CI changes for Ubuntu 24

nmoinvaz opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
	- Updated GitHub Actions workflows for multiple repositories.
	- Standardized package management for cross-compilation environments.
	- Upgraded Clang compiler versions from 11 to 15.
	- Simplified package dependencies by removing redundant `qemu` packages.
	- Updated cross-compilation package names for various architectures.
	- Renamed job configurations for clarity and consistency.
	- Updated LLVM package versions across various workflows.
	- Enhanced consistency in coverage reporting across multiple job configurations.
	- Added `gcov-exec` specifications for improved coverage reporting.
	- Modified variable types in macros to accommodate larger input data sizes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nmoinvaz committed:
Upgrade CI from Clang-11 to Clang 15 for Ubuntu 24. #1840

nmoinvaz committed:
Don't pin gcovr version any longer. #1840

nmoinvaz committed:
Remove package qemu for Ubuntu 24. #1840

nmoinvaz committed:
Ignore gcovr parser errors.

nmoinvaz commented:
We are getting some illegal instructions on Ubuntu GCC PCC64LE:
```
qemu-ppc64le;-cpu;power8;-L;/usr/powerpc64le-linux-gnu/;/home/runner/work/zlib-ng/zlib-ng/minigzip;-k;-c;-2
18:   failed: Illegal instruction
```

nmoinvaz committed:
Use Ubuntu 24 crossbuild-essential packages.

Dead2 commented:
> We are getting some illegal instructions on Ubuntu GCC PCC64LE:
> 
> ```
> qemu-ppc64le;-cpu;power8;-L;/usr/powerpc64le-linux-gnu/;/home/runner/work/zlib-ng/zlib-ng/minigzip;-k;-c;-2
> 18:   failed: Illegal instruction
> ```

If I understand the CI and toolchain differences correctly, the failure happens only with the Power8 + Little Endian + VSX combination.
GCC fails on Power8 LE with VSX.
GCC succeeds on Power8 LE without VSX.
GCC succeeds on Power8 BE with VSX.
Clang does not seem to be tested with Power8, only Power9.

So I am guessing something might be wrong with Power8 VSX code, somehow affected by LE endianness.
Maybe @KungFuJesus has some more insight.

nmoinvaz committed:
Use correct version of gcov for cross-compilers.

nmoinvaz commented:
I commented out all the pc*/power* assigns in functable.h and errors still happen.

nmoinvaz commented:
We need to ignore gcov errors. Seems they were previously ignored anyways.
https://github.com/linux-test-project/lcov/issues/341#issuecomment-2548282709

nmoinvaz committed:
Add missing compiler-rt libraries for Ubuntu 24. #1840

nmoinvaz commented:
[Ubuntu GCC AARCH64](https://github.com/nmoinvaz/zlib-ng/actions/runs/12620816900/job/35166940379#logs) builds failing due to segmentation fault when running gtest_zlib. Perhaps that was mentioned somewhere else?

nmoinvaz commented:
I have a Raspberry Pi 5 with Ubuntu 24.10 and I was able to run gtest_zlib after compiling with both clang 19 and gcc 14. So perhaps it is an issue with qemu. I'm inclined to just force those tests to use Ubuntu 22.

Dead2 commented:
> I have a Raspberry Pi 5 with Ubuntu 24.10 and I was able to run gtest_zlib after compiling with both clang 19 and gcc 14. So perhaps it is an issue with qemu. I'm inclined to just force those tests to use Ubuntu 22.

Yeah, how about we have those do Ubuntu 22 and just add an issue about it being broken, that way we might remember it and could recheck it at a later time, or someone else might take it upon themselves to investigate it further.

nmoinvaz committed:
Use Ubuntu 22.04 for AARCH64 tests

It seems that qemu might be failing. Tests on Raspberry Pi 5 with Ubuntu 24.04
appear to work just fine.

mtl1979 commented:
> We are getting some illegal instructions on Ubuntu GCC PCC64LE:
> 
> ```
> qemu-ppc64le;-cpu;power8;-L;/usr/powerpc64le-linux-gnu/;/home/runner/work/zlib-ng/zlib-ng/minigzip;-k;-c;-2
> 18:   failed: Illegal instruction
> ```

Support for Power8 was dropped, you need older version of toolchain. This was mentioned earlier.

nmoinvaz commented:
Where was it mentioned?

nmoinvaz commented:
I found it.. https://github.com/zlib-ng/zlib-ng/issues/1378

nmoinvaz commented:
Is it just me or does qemu appear to be getting worse and worse?

nmoinvaz commented:
MSAN fails because zlib-ng repo isn’t updated to use Ubuntu 24 as ubuntu-latest yet. In my branch it works.

mtl1979 commented:
> Is it just me or does qemu appear to be getting worse and worse?

It has been broken since around 6.0 and worse performance wise since they ditched the old instruction translator that was based on old GCC.

mtl1979 commented:
> MSAN fails because zlib-ng repo isn’t updated to use Ubuntu 24 as ubuntu-latest yet. In my branch it works.

In most cases it is better to use explicit version instead of "latest" as during deploying of new Ubuntu release every 2 years, a lot of package versions change and we rely on specific matching package versions for certain package combinations.

nmoinvaz commented:
Rebased

Dead2 commented:
Still some failed runs, did not spot anything obvious, other than the Power8 issue you discussed above.

nmoinvaz commented:
Looks like my commit for PPC64LE was overwritten on rebase.

nmoinvaz committed:
Use Ubuntu 20.04 for PPC64LE tests due to broken qemu.

nmoinvaz commented:
Not sure about the "Package check" ones.

phprus commented on inflate.c:
```diff
@@ -387,7 +387,7 @@ static void updatewindow(PREFIX3(stream) *strm, const uint8_t *end, uint32_t len
         if (have == 0) goto inf_leave; \
         have--; \
-        hold += ((unsigned)(*next++) << bits); \
+        hold += ((uint64_t)(*next++) << bits); \
```
Is this the correct fix?
The ``hold`` variable is of type ``uint32_t`` and may overflow when assigned with addition.

nmoinvaz replied:
`hold` is `uint64_t`. The warning goes away, so I assume it is correct.

nmoinvaz replied:
There are two separate warnings for two different lines. I think they are non-issues, and AI suggested something that looks rather complicated.

phprus replied:
> `hold` is `uint64_t`. The warning goes away, so I assume it is correct.

https://github.com/zlib-ng/zlib-ng/blob/8ff036958987f2acbc92daaca32066ade20e63b0/inflate.c#L482

https://github.com/zlib-ng/zlib-ng/blob/8ff036958987f2acbc92daaca32066ade20e63b0/infback.c#L157

??

nmoinvaz replied:
It looks like it is `uint64_t` in `inffast_tpl.h`.

phprus replied:
Potentially dangerous:

``ZSWAP32(hold)``

on this line and below:

https://github.com/zlib-ng/zlib-ng/blob/e6fca08c04c82e03c802ca6107de9e70abd64a0a/inflate.c#L707

nmoinvaz replied:
Hmm. I think I will just let somebody else fix this static analysis problem.

Dead2 reviewed (approved):
LGTM.
While this does not fix all the current failures, it makes it actually possible to see the rest of the failures instead of them drowning in a sea of other failures 👍
